In [1]:
from seq_utils import read_fasta

In [2]:
filepath = '../../dataset/Homo_sapiens.GRCh38.cdna.all.fa'
cdnas = read_fasta(filepath)

In [3]:
cdnas[:2]

[{'transcript_id': 'ENST00000415118.1',
  'attributes': {'transcript_id': 'ENST00000415118.1',
   'seqtype': 'cdna',
   'chromosome': 'GRCh38:14:22438547:22438554:1 ',
   'gene': 'ENSG00000223997.1 ',
   'gene_biotype': 'TR_D_gene ',
   'transcript_biotype': 'TR_D_gene ',
   'gene_symbol': 'TRDD1 ',
   'description': 'T cell receptor delta diversity 1 [Source:HGNC Symbol;Acc:HGNC:12254]'},
  'sequence': 'GAAATAGT'},
 {'transcript_id': 'ENST00000448914.1',
  'attributes': {'transcript_id': 'ENST00000448914.1',
   'seqtype': 'cdna',
   'chromosome': 'GRCh38:14:22449113:22449125:1 ',
   'gene': 'ENSG00000228985.1 ',
   'gene_biotype': 'TR_D_gene ',
   'transcript_biotype': 'TR_D_gene ',
   'gene_symbol': 'TRDD3 ',
   'description': 'T cell receptor delta diversity 3 [Source:HGNC Symbol;Acc:HGNC:12256]'},
  'sequence': 'ACTGGGGGATACG'}]

In [77]:
filepath = '../../dataset/Homo_sapiens.GRCh38.cds.all.fa'
cdss = read_fasta(filepath)

In [86]:
cdss[:2]

[{'transcript_id': 'ENST00000415118.1',
  'attributes': {'transcript_id': 'ENST00000415118.1',
   'seqtype': 'cds',
   'chromosome': 'GRCh38:14:22438547:22438554:1 ',
   'gene': 'ENSG00000223997.1 ',
   'gene_biotype': 'TR_D_gene ',
   'transcript_biotype': 'TR_D_gene ',
   'gene_symbol': 'TRDD1 ',
   'description': 'T cell receptor delta diversity 1 [Source:HGNC Symbol;Acc:HGNC:12254]'},
  'sequence': 'GAAATAGT'},
 {'transcript_id': 'ENST00000448914.1',
  'attributes': {'transcript_id': 'ENST00000448914.1',
   'seqtype': 'cds',
   'chromosome': 'GRCh38:14:22449113:22449125:1 ',
   'gene': 'ENSG00000228985.1 ',
   'gene_biotype': 'TR_D_gene ',
   'transcript_biotype': 'TR_D_gene ',
   'gene_symbol': 'TRDD3 ',
   'description': 'T cell receptor delta diversity 3 [Source:HGNC Symbol;Acc:HGNC:12256]'},
  'sequence': 'ACTGGGGGATACG'}]

In [4]:
from seq_utils import read_fastq_barcode
seq_path = './data/500_PBMC_3p_LT_Chromium_X_S4_L004_R2_001.fastq_ds100000'
bc_path = './data/500_PBMC_3p_LT_Chromium_X_S4_L004_R1_001.fastq_ds100000'
sequences = read_fastq_barcode(seq_path, bc_path)

In [5]:
sequences[:10]

[{'barcode': 'GNGGGTAAGTGTTGAAGGATTTTTATGG',
  'attributes': {'instrument': 'A00519',
   'run_number': '986',
   'flowcell_ID': 'HFFLJDSX2',
   'lane': '4',
   'tile': '1101',
   'x_pos': '15510',
   'y_pos': '1016',
   'read': '2',
   'is_filtered': 'N',
   'control number': '0',
   'i7_index': 'GAGACGCACG',
   'i5_index': 'ATGTTCATAG'},
  'sequence': 'AAGCAGTGGTATCAACGCAGAGTACATGGGGCTATGCTGGAGGTTTATATGAAATACAGATAAAGCACACAAAGATTGAGTCTATGACAA',
  'quality': 'F:FFF:,,FFF,FFFFFF:,FF:FF:FF,FFF,F,F,F:FFFF:FFF:,,F:F:FFFFFFF:FFFFFFFFF,FF:,:FFFFFF,FF:FF:',
  'bc_quality': 'F#F:FFF,:,FF,FF,FFFFFF,F::,:'},
 {'barcode': 'ANTGACTTCGTTCCTGTAAACTGATTCT',
  'attributes': {'instrument': 'A00519',
   'run_number': '986',
   'flowcell_ID': 'HFFLJDSX2',
   'lane': '4',
   'tile': '1101',
   'x_pos': '16089',
   'y_pos': '1016',
   'read': '2',
   'is_filtered': 'N',
   'control number': '0',
   'i7_index': 'GAGACGCACG',
   'i5_index': 'ATGTTCATAG'},
  'sequence': 'GGGGGATAGAAAAGAAATAAGCAGGCCAGGCTCAGTGGC

In [16]:
from seq_utils import smith_waterman, substitution_matrix, gap_penalty

In [6]:
temp = [i for i in cdnas if i['transcript_id'] == 'ENST00000488803.1'][0]

 {'barcode': 'ANTAGTCTCAAAGGTAACCGCTCTTGTA',  
  'attributes': {'instrument': 'A00519',  
   'run_number': '986',  
   'flowcell_ID': 'HFFLJDSX2',  
   'lane': '4',  
   'tile': '1101',  
   'x_pos': '28076',  
   'y_pos': '1016',  
   'read': '2',  
   'is_filtered': 'N',  
   'control number': '0',  
   'i7_index': 'GAGACGCACG',  
   'i5_index': 'ATGTTCATAG'},  
  'sequence': 'CTGCTCATGATGGCTGGTATCGATGACTGCTACACCTCAGCCCGGGGCTGCACTGCCACCCTGGGCAACTTCGCCAAGGCCACCTTTGAT',  
  'quality': 'FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF:FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF:FFF::',  
  'bc_quality': 'F#FFFFFFFFFFFFF:FFFFFFFFFFFF'},

In [83]:
smith_waterman('CTGCTCATGATGGCTGGTATCGATGACTGCTACACCTCAGCCCGGGGCTGCACTGCCACCCTGGGCAACTTCGCCAAGGCCACCTTTGAT', 
               temp['sequence'], substitution_matrix, gap_penalty)


(180.0,
 'CTGCTCATGATGGCTGGTATCGATGACTGCTACACCTCAGCCCGGGGCTGCACTGCCACCCTGGGCAACTTCGCCAAGGCCACCTTTGAT',
 'CTGCTCATGATGGCTGGTATCGATGACTGCTACACCTCAGCCCGGGGCTGCACTGCCACCCTGGGCAACTTCGCCAAGGCCACCTTTGAT')

In [8]:
temp = [i for i in cdnas if i['transcript_id'] == 'ENST00000467001.1'][0]

 {'barcode': 'TNTCGATAGTGCGTCCTAATAAACGGGT',  
  'attributes': {'instrument': 'A00519',  
   'run_number': '986',  
   'flowcell_ID': 'HFFLJDSX2',  
   'lane': '4',  
   'tile': '1101',  
   'x_pos': '27932',  
   'y_pos': '1016',  
   'read': '2',  
   'is_filtered': 'N',  
   'control number': '0',  
   'i7_index': 'GAGACGCACG',  
   'i5_index': 'ATGTTCATAG'},  
  'sequence': 'GCAGGACTTCCTCAGTGCCCAGCAGGTGCAGGCCCCTGTGAAGCTCTATTCTGACTGGCTGTCCGTGGGCCACGTGGACGAGTTCCTGAG',  
  'quality': 'FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF:FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF:F',  
  'bc_quality': 'F#FFFFFFFFFFFFFFFFFFFFFFFFFF'},

In [85]:
smith_waterman('GCAGGACTTCCTCAGTGCCCAGCAGGTGCAGGCCCCTGTGAAGCTCTATTCTGACTGGCTGTCCGTGGGCCACGTGGACGAGTTCCTGAG', 
               temp['sequence'], substitution_matrix, gap_penalty)

(180.0,
 'GCAGGACTTCCTCAGTGCCCAGCAGGTGCAGGCCCCTGTGAAGCTCTATTCTGACTGGCTGTCCGTGGGCCACGTGGACGAGTTCCTGAG',
 'GCAGGACTTCCTCAGTGCCCAGCAGGTGCAGGCCCCTGTGAAGCTCTATTCTGACTGGCTGTCCGTGGGCCACGTGGACGAGTTCCTGAG')

In [19]:
temp = [i for i in cdnas if i['transcript_id'] == 'ENST00000495554.1'][0]

{'barcode': 'GNGTTGTCAGACCTATGCGGATCAGTAT',  
  'attributes': {'instrument': 'A00519',  
   'run_number': '986',  
   'flowcell_ID': 'HFFLJDSX2',  
   'lane': '4',  
   'tile': '1101',  
   'x_pos': '20518',  
   'y_pos': '1016',  
   'read': '2',  
   'is_filtered': 'N',  
   'control number': '0',  
   'i7_index': 'GAGACGCACG',  
   'i5_index': 'ATGTTCATAG'},  
  'sequence': 'CAGTGGTATCAACGCAGAGTACATGGGAAAGAAAAGTAAAGATATTCAAAGGCAGTTTCTATCTCTGACAACTTTGTTCATTAAGAGTGC',  
  'quality': 'FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF:F,FFFFFFFFFFFFFFFFFFFFFFF:FFFFFFFFFFFFFFFFFFFFFFFF',  
  'bc_quality': 'F#FFFFFFFFFFFFFFFFFFFFFFFFFF'},

In [20]:
smith_waterman('CAGTGGTATCAACGCAGAGTACATGGGAAAGAAAAGTAAAGATATTCAAAGGCAGTTTCTATCTCTGACAACTTTGTTCATTAAGAGTGC', 
               temp['sequence'], substitution_matrix, gap_penalty)

(137.0,
 'TGGTATCA-ACGCAGAG-TACATGG-GA--AAGAAAAGTAAAGATATTCAAAGGCAGTTTCTATCTCTGACAACTTTGTTCATTAAGAGTGC',
 'TGGAGTCAGAAAAAGTGATGATTGGAGACTAAGAAAAGTAAAGATATTCAAAGGCAGTTTCTATCTCTGACAACTTTGTTCATTAAGAGTGC')

In [21]:
temp = [i for i in cdnas if i['transcript_id'] == 'ENST00000486372.1'][0]
temp

{'transcript_id': 'ENST00000486372.1',
 'attributes': {'transcript_id': 'ENST00000486372.1',
  'seqtype': 'cdna',
  'chromosome': 'GRCh38:9:129803157:129804480:1 ',
  'gene': 'ENSG00000136816.16 ',
  'gene_biotype': 'protein_coding ',
  'transcript_biotype': 'protein_coding_CDS_not_defined ',
  'gene_symbol': 'TOR1B ',
  'description': 'torsin family 1 member B [Source:HGNC Symbol;Acc:HGNC:11995]'},
 'sequence': 'GCCGCTTGGCCGCGGGGCGCCTGGCTCAGTGGCTTCTGCGGGCTTCGAGGAGCGGGATGTTGCGGGCTGGGTGGCTCCGGGGCGCGGCGGCGCTGGCGCTGCTGCTGGCGGCCCGAGTGGTGGCGGCGTTCGAGCCCATCACCGTGGGCCTAGCCATCGGGGCCGCGTCGGCCATCACCGGCTACCTGTCCTACAATGACATCTACTGCCGCTTCGCCGAGTGCTGCCGCGAGGAGCGGCCGCTCAACGCTTCGGCTCTCAAGCTGGATTTGGAGGAGAAGCTGTTTGGACAGCATCTAGCCACGGAAGTGATTTTCAAGGCGCTGACTGGCTTCAGGAACAACAAAAATCCCAAGAAACCACTGACCCTTTCCTTACACGGCTGGGCTGGCACAGGCAAGAATTTTGTCAGTCAAATTGTGGCTGAAAATCTTCACCCAAAAGGTCTGAAGAGTAACTTTGTCCACCTGTTTGTATCGACTCTGCACTTCCCTCATGAGCAGAAGATAAAACTGTACCAGGCAAGAGAACCCGCTATTATCTCGTCTGCAGGCCAGTCGGACTGGTCCGGGTGACCTGCTCA

In [22]:
smith_waterman('AAAAATCCCAAGAAACCACTGACCCTTTCCTTACACGGCTGGGCTGGCACAGGCAAGAATTTTGTCAGTCAAATTTTGGCTGAAAATCTT', 
               temp['sequence'], substitution_matrix, gap_penalty)

(177.0,
 'AAAAATCCCAAGAAACCACTGACCCTTTCCTTACACGGCTGGGCTGGCACAGGCAAGAATTTTGTCAGTCAAATTTTGGCTGAAAATCTT',
 'AAAAATCCCAAGAAACCACTGACCCTTTCCTTACACGGCTGGGCTGGCACAGGCAAGAATTTTGTCAGTCAAATTGTGGCTGAAAATCTT')

In [9]:
def reverse_complement(dna_sequence):
    """
    Returns the reverse complement of a given DNA sequence.
    
    Args:
    dna_sequence: A string representing the DNA sequence.
    
    Returns:
    A string representing the reverse complement of the DNA sequence.
    """
    # Create a dictionary to map each nucleotide to its complement
    complement = {'A': 'T', 'T': 'A', 'C': 'G', 'G': 'C', 'N': 'N'}
    
    # Generate the complement sequence
    complement_sequence = ''.join(complement[base] for base in dna_sequence)
    
    # Reverse the complement sequence
    reverse_complement_sequence = complement_sequence[::-1]
    
    return reverse_complement_sequence

# Example usage:
dna_sequence = "AGTACGCATACGGCATA"
print(reverse_complement(dna_sequence))  # Output: TATGCCGTATGCGTACT


TATGCCGTATGCGTACT
